node/narwhal/tests/common/utils.rs

```diff
@@ -0,0 +1,148 @@
+
+use snarkos_node_narwhal::helpers::PrimarySender;
+use snarkvm::{
+    ledger::narwhal::Data,
+    prelude::{
+        block::Transaction,
+        coinbase::{ProverSolution, PuzzleCommitment},
+        Field,
+        Network,
+        Uniform,
+    },
+};
+use tracing::*;
+
+use ::bytes::Bytes;
+
+use rand::{Rng, SeedableRng};
+
+use tracing_subscriber::{
+    layer::{Layer, SubscriberExt},
+    util::SubscriberInitExt,
+};
+
+use crate::common::CurrentNetwork;
+
+/// Initializes the logger.
+#[allow(dead_code)]
+pub fn initialize_logger(verbosity: u8) {
+    match verbosity {
+        0 => std::env::set_var("RUST_LOG", "info"),
+        1 => std::env::set_var("RUST_LOG", "debug"),
+        2..=4 => std::env::set_var("RUST_LOG", "trace"),
+        _ => std::env::set_var("RUST_LOG", "info"),
+    };
+
+    // Filter out undesirable logs. (unfortunately EnvFilter cannot be cloned)
+    let [filter] = std::array::from_fn(|_| {
+        let filter = tracing_subscriber::EnvFilter::from_default_env()
+            .add_directive("mio=off".parse().unwrap())
+            .add_directive("tokio_util=off".parse().unwrap())
+            .add_directive("hyper=off".parse().unwrap())
+            .add_directive("reqwest=off".parse().unwrap())
+            .add_directive("want=off".parse().unwrap())
+            .add_directive("warp=off".parse().unwrap());
+
+        if verbosity > 3 {
+            filter.add_directive("snarkos_node_tcp=trace".parse().unwrap())
+        } else {
+            filter.add_directive("snarkos_node_tcp=off".parse().unwrap())
+        }
+    });
+
+    // Initialize tracing.
+    let _ = tracing_subscriber::registry()
+        .with(tracing_subscriber::fmt::Layer::default().with_target(verbosity > 2).with_filter(filter))
+        .try_init();
+}
+
+/// Fires *fake* unconfirmed solutions at the node.
+pub fn fire_unconfirmed_solutions(sender: &PrimarySender<CurrentNetwork>, node_id: u16) {
+    let tx_unconfirmed_solution = sender.tx_unconfirmed_solution.clone();
+    tokio::task::spawn(async move {
+        // This RNG samples the *same* fake solutions for all nodes.
+        let mut shared_rng = rand_chacha::ChaChaRng::seed_from_u64(123456789);
+        // This RNG samples *different* fake solutions for each node.
+        let mut unique_rng = rand_chacha::ChaChaRng::seed_from_u64(node_id as u64);
+
+        // A closure to generate a commitment and solution.
+        fn sample(mut rng: impl Rng) -> (PuzzleCommitment<CurrentNetwork>, Data<ProverSolution<CurrentNetwork>>) {
+            // Sample a random fake puzzle commitment.
+            // TODO (howardwu): Use a mutex to bring in the real 'proof target' and change this sampling to a while loop.
+            let commitment = PuzzleCommitment::<CurrentNetwork>::from_g1_affine(rng.gen());
+            // Sample random fake solution bytes.
+            let solution = Data::Buffer(Bytes::from((0..1024).map(|_| rng.gen::<u8>()).collect::<Vec<_>>()));
```

Comment:
instead of `gen`, `try_fill` would be more efficient; it would look sth like this:
```
let mut vec = vec![0u8; 1024];
vec.try_fill(&mut rng).unwrap();
let solution = Data::Buffer(Bytes::from(vec));
```

Reply:
done, used `fill_bytes` in the end